decoder/correlator.go

```diff
@@ -12,140 +12,309 @@ import (
-				restCallID := payload[posCallID:]
-				// Minimum Call-ID length of "Call-ID: a" = 10
-				if posRestCallID := bytes.Index(restCallID, []byte("\r\n")); posRestCallID >= 10 {
-					callID = restCallID[len("Call-ID:"):posRestCallID]
-				} else {
-					logp.Debug("sdp", "No end or fishy Call-ID in '%s'", restCallID)
-					return
+			// Reset rtcp data for this media.
+			rtcpIP = sessionIP
+			rtcpPort = nil
+			// We are only interested in audio.
+			if !bytes.HasPrefix(line, []byte("m=audio ")) {
+				continue sdpLoop
+			}
+			// Find separator after rtp port number.
+			sep := bytes.Index(line[8:], []byte(" "))
+			if sep < 4 { // Port should be above 1000
+				logp.Debug("sdp", "Fishy m=audio line %q. callID=%q", line, callID)
+				continue sdpLoop
+			}
+			// Extract rtp port.
+			rtpPort := line[8 : 8+sep]
+			// Check for and strip port count.
+			sep2 := bytes.Index(rtpPort, []byte("/"))
+			if sep2 > 0 {
+				rtpPort = rtpPort[:sep2]
+			}
+			// Convert from rtp port to rtcp port by adding 1.
+			// Do not assume that rtp port is even.
+			rtpPortNb, err2 := strconv.Atoi(string(rtpPort))
+			if err2 != nil {
+				logp.Debug("sdp", "Fishy m=audio line %q. callID=%q", line, callID)
+				continue sdpLoop
+			}
+			rtcpPort = []byte(strconv.Itoa(rtpPortNb + 1))
+		case 'a':
+			// We are only interested in a=rtcp.
+			if !bytes.HasPrefix(line, []byte("a=rtcp:")) {
+				continue sdpLoop
+			}
+			// May contain only port or port and ip.
+			sep := bytes.Index(line[7:], []byte(" "))
+			if sep < 0 {
+				// Port only.
+				rtcpPort = line[7:]
+			} else {
+				// Port and ip, e.g. "1000 IN IP4 1.1.1.1".
+				if !bytes.HasPrefix(line[7+sep+1:], []byte("IN IP")) {
+					logp.Debug("sdp", "Fishy a=rtcp line %q. callID=%q", line, callID)
+					continue sdpLoop
 				}
-			} else if posID := bytes.Index(payload, []byte("i: ")); posID > 0 {
-				restID := payload[posID:]
-				// Minimum Call-ID length of "i: a" = 4
-				if posRestID := bytes.Index(restID, []byte("\r\n")); posRestID >= 4 {
-					callID = restID[len("i: "):posRestID]
-				} else {
-					logp.Debug("sdp", "No end or fishy Call-ID in '%s'", restID)
-					return
+				// Extract port.
+				rtcpPort = line[7 : 7+sep]
+				// Check for and strip count.
+				sep2 := bytes.Index(rtcpPort, []byte("/"))
+				if sep2 > 0 {
+					rtcpPort = rtcpPort[:sep2]
+				}
+				// Extract ip.
+				rtcpIP = line[7+sep+1+5+2:] // space + "IN IP" + version + space.
+				// Check for and strip ttl/count separated by slash.
+				sep3 := bytes.Index(rtcpIP, []byte("/"))
+				if sep3 > 0 {
+					rtcpIP = rtcpIP[:sep3]
 				}
-			} else {
-				logp.Warn("No Call-ID in '%s'", payload)
-				return
 			}
-
-			//logp.Debug("sdp", "Add to SDPCache key=%s, value=%s", ipPort.String(), string(callID))
-			sdpCache.Set(ipPort.Bytes(), bytes.TrimSpace(callID), 1200)
+		default:
+			// ignore other sdp lines.
+		}
+	}
+	// Add keys for last media.
+	if len(rtcpIP) > 0 && len(rtcpPort) > 0 {
```

Comment:
Mby we should add a comment why you choose this solution over for example this one:

```
	if len(rtcpIP) > 0 && len(rtcpPort) > 0 {
		if bytes.Equal(rtcpIP, srcIPb) {
			addSDPCacheEntry(rtcpIP, rtcpPort, callID)
		} else {
			// rtcp ip may be natted, use seen source ip too.
			addSDPCacheEntry(srcIPb, rtcpPort, callID)
		}
	}
```

Reply:
I assume you mean the code from 227 to 234 and 156 to 163.
I use both ips, as one can not know which ip is the right one. If there is no nat or the sender corrected the sdp and an extra ip was given in a=rtcp than the ips will differ but the rtcpIP is correct. If there is nat and sdp was not corrected by sender, than srcIP probably works.
The equal check is just for speedup. Most times both ips will be se same and adding the same key twice is probably slower than doing an simple extra check.

Reply:
Yeah I wanted to mark multiple lines but only selected one ;) Ok mby we can add a comment. Because it may not be clear for everyone why in case of rtcpIP!=srcIPb you basicly do this:

addSDPCacheEntry(rtcpIP, rtcpPort, callID)
addSDPCacheEntry(srcIPb, rtcpPort, callID)

Reply:
Makes sense, I will do that.